### PR TITLE
Defer support

### PR DESCRIPTION
**Pull request by go-wyvern**
have some defer feature should to test，commit pr first.

### CODE REVIEW

**Comment by JessonChan** [comment]
1、defer 要满足后进先出特点，你的实现在block和return两个地方进行defer的编译这点无法保证。
2、defer也是普通语句，可以用到任何地方，也不只有block中。比如
```go
package main

import "fmt"

func main() {
	func() {
		for i := 0; i < 10; i++ {
			defer fmt.Println(i)
		}
	}()
}
```

**Comment by go-wyvern** [comment]
> 1、defer 要满足后进先出特点，你的实现在block和return两个地方进行defer的编译这点无法保证。
> 2、defer也是普通语句，可以用到任何地方，也不只有block中。比如
> 
> ```go
> package main
> 
> import "fmt"
> 
> func main() {
> 	func() {
> 		for i := 0; i < 10; i++ {
> 			defer fmt.Println(i)
> 		}
> 	}()
> }
> ```

thanks for your comments.  
I will carefully read the code in the compile part, which is not very familiar

**Comment by xushiwei** [comment]
You may can refer `defer` of qlang:
* https://github.com/xushiwei/qlang/blob/v1.5/exec/code.go#L287
* https://github.com/xushiwei/qlang/blob/v1.5/exec/code.go#L345
* https://github.com/xushiwei/qlang/blob/v1.5/exec/function.go#L85
* https://github.com/xushiwei/qlang/blob/v1.5/cl/function2.go#L56

However, there are some differences between golang's `defer` and qlang's. For example:

```go
f, err := os.Open(...)
if err != nil {
   return
}
defer f.Close()

use(f)
f = nil // NOTE!
```

In qlang, `f.Close()` will be executed when function exists. The value of variable `f` is got until that time. So the above code will crash.

In go, `f.Close()` will be executed when function exists, but all the values of variables are got before adding `f.Close()` to `defer` chain of the function.

**Comment by xushiwei** [comment]
And let's see another example (https://play.golang.org/p/BbjlrYxe126):

```go
package main

import (
	"fmt"
)

func main() {
	defer fmt.Println(fmt.Println("Hello, playground"))
	fmt.Println("start")
}
```

The output of go is:

```
Hello, playground
start
18 <nil>
```

And then you can know that `defer` chain is only a simple function calling chain.

**Comment by go-wyvern** [comment]
> You may can refer `defer` of qlang:
> 
> * https://github.com/xushiwei/qlang/blob/v1.5/exec/code.go#L287
> * https://github.com/xushiwei/qlang/blob/v1.5/exec/code.go#L345
> * https://github.com/xushiwei/qlang/blob/v1.5/exec/function.go#L85
> * https://github.com/xushiwei/qlang/blob/v1.5/cl/function2.go#L56
> 
> However, there are some differences between golang's `defer` and qlang's. For example:
> 
> ```go
> f, err := os.Open(...)
> if err != nil {
>    return
> }
> defer f.Close()
> 
> use(f)
> f = nil // NOTE!
> ```
> 
> In qlang, `f.Close()` will be executed when function exists. The value of variable `f` is got until that time. So the above code will crash.
> 
> In go, `f.Close()` will be executed when function exists, but all the values of variables are got before adding `f.Close()` to `defer` chain of the function.

thanks for your comments.

Your comments have inspired me a lot, I have made changes to my code, please review

File: exec/bytecode/defer.go [hunk 1]
```diff
@@ -0,0 +1,78 @@
...
+// Package bytecode implements a bytecode backend for the Go+ language.
+package bytecode
+
+import (
+	"github.com/goplus/gop/exec.spec"
+)
+
+type theDefer struct {
+	next *theDefer
+	Stack
```
Comment (xushiwei) [inline exec/bytecode/defer.go:26]: why need a stack?

Comment (go-wyvern) [reply exec/bytecode/defer.go:26]: Save the return value of the function in the defer function parameter.  example:

```
package main

import (
	"fmt"
)

func main() {
	defer fmt.Println(fmt.Println("Hello, playground"))
	fmt.Println("start")
}
```

Saved stack output:

```
Hello, playground
start
18 <nil>
```

The stack output is not saved：

```
Hello, playground
start
7 <nil>
```

Comment (xushiwei) [reply exec/bytecode/defer.go:26]: I see. It really works, but may have a better solution.
I will merge it at first.

File: exec/golang/stmt.go [hunk 2]
```diff
@@ -393,4 +393,9 @@ func (p *Builder) EndComprehension(c *Comprehension) *Builder {
 	return p
 }
 
+func (p *Builder) Defer(start, end *Label) exec.Instr {
+	// TODO
```
Comment (xushiwei) [inline exec/golang/stmt.go:397]: `panic` will be better.

File: exec/bytecode/defer.go (hunk 1 above)
Comment (go-wyvern) [reply exec/bytecode/defer.go:26]: Ok, look forward to a better solution

File: exec/golang/stmt.go (hunk 2 above)
Comment (go-wyvern) [reply exec/golang/stmt.go]: Ok, modified

**Comment by xushiwei** [comment]
code cover is still not enough.

**Comment by go-wyvern** [comment]
Okay, I will add some test cases later

**Comment by go-wyvern** [comment]
> code cover is still not enough.

I have added some unit tests, and now the code cover has successfully run

**Pull request merged**